pricegraph/src/graph/shortest_paths.rs

```diff
@@ -11,25 +11,53 @@ use petgraph::visit::{
 /// A vector associating to each node index a distance from the source.
 type Distance<G> = Vec<<G as Data>::EdgeWeight>;
 
+type PredecessorVec<G> = Vec<Option<<G as GraphBase>::NodeId>>;
+
+struct UpdatableDistance<G: Data> {
+    current: Distance<G>,
+    update: Distance<G>,
+}
+
 /// Stores the information needed to manage the predecessor list.
-/// For each node index, this type contains its predecessor node and
-/// distance in the graph from a source node.
-struct PredecessorStore<G: GraphBase + Data> {
-    predecessors: Vec<Option<G::NodeId>>,
-    distance: Distance<G>,
+/// For each node index, this type contains the information needed to
+/// compute predecessor nodes and the distance in the graph from a source
+/// node.
+enum PredecessorStore<G: GraphBase + Data> {
+    Unbounded(PredecessorVec<G>, Distance<G>),
+    Bounded(Vec<PredecessorVec<G>>, UpdatableDistance<G>),
 }
```

Comment:
A lot of functions in this file have to match on this and do different things based on the case. While reading this I was wondering if there was a more readable way to implement this by having each algorithm (no hops vs hops) in its own file and connecting them in a third file either with match or through a trait.
However, it is not clear what the best approach or even if there is a better one than this so I think it's okay if we do that after merging the hops related PRs if we still feel the need then.

Reply:
I think this is a great suggestion. It does look like the algorithms are quite different so maybe implementing them in separate traits is a good idea.

Reply:
> separate traits

I agree minus that we should use one trait and separate implementations.